### ISSUE TITLE
Secrets and Encryption

### ISSUE THREAD

**Issue by laardee** [serverless/components#238]
Here are some thoughts about secrets and encryption that could be implemented to the Components. With encrypted state.json and encrypted inline secrets, working with version control would be easier.

By default, encryption could use Node.js native crypto module with `aes-256-cbc` algorithm. Optionally user could define AWS KMS, GCP KMS or similar service to handle the encryption and key. These optional methods could be implemented to the core or some kind of plugins or extensions. 

## Encrypted state.json

With command `components encrypt --state` the state file is encrypted, and after that, it is encrypted every time before written to disk or other storage. 

**example encrypted state.json**
```json
{
  "encrypted": "ckE+aBv0Ja9QK+GuvVMiq4+5O0OEk9+LFR1VBK+OUS4CXh02gFOJZyKF/qCFBOUcUYV6ho6ontyoQFBED6SjUbMywnS+gZ2wtyq7XMMzMhUVjtPMEO6cbalT4SRImXe5J7S0g66XnrFmIyMB8JWbkj3uqUvZgcXLSBOdAbHZZdx7z/ktXKAxtGjCu9NWK8eNFKpVV+5osnWqVMAoezTOYg=="
}
```

`components decrypt --state`  would do the opposite.

## Inline variables in serverless.yml

Using encrypted values in serverless.yml would allow it to be committed to version control with inline secrets.

Running `components encrypt --variable some-secret-value` would output the encrypted value e.g. `E5oQO0xd3nmkk/tKuaenjrMKS3XmCJIKa+wVgfzU7kw=`. That value then can be used in serverless.yml as `!encrypted` type.

```yaml
type: my-service

inputs:
  secretToken: !encrypted |
    E5oQO0xd3nmkk/tKuaenjrMKS3XmCJIKa+wVgfzU7kw=
```

`!encrypted` type is implemented as a custom `js-yaml` type + schema in the Components codebase. The schema is passed to @serverless/utils readFile->parseFile, so that the decryption is done while parsing the yaml data to js object.

`components decrypt --variable E5oQO0xd3nmkk/tKuaenjrMKS3XmCJIKa+wVgfzU7kw=` decrypts the value and displays it as plain text.

There could be also an option to encrypt values that are defined in serverless.yml, e.g. `components encrypt --path inputs.secretToken` would replace the value of
```yaml
inputs:
  secretToken: some-secret-value
```
with
```yaml
inputs:
  secretToken: !encrypted E5oQO0xd3nmkk/tKuaenjrMKS3XmCJIKa+wVgfzU7kw=
```

Any thoughts?

@brianneisler  @eahefnawy  @pmuens  @ac360

**Comment by eahefnawy** [comment]
@laardee this is an old issue, but how do you think this could work with the new implementation?

**Comment by laardee** [comment]
@eahefnawy yeah, I think this or something similar would be worth to implement.

**Comment by eahefnawy** [comment]
@laardee do you think remote state storage could solve this? If so, I'd like to close this issue 😊
